### DISCUSSION TITLE
[Android] How to exclude resource files from being copied to output on build

### DISCUSSION THREAD

**Discussion by rh101** [axmolengine/axmol#974] [Q&A]
Let's say the game resource folder contains the following files:

file1.png
file2.mp3
file3.pem

On build, "file3.pem" should not be copied to the output folder, just prior to the APK being created.

The documentation I've seen online indicates that I have to do add `exclude` lines in this section:

In `proj.andoid/app/build.gralde`

```
android.applicationVariants.all { variant ->
...    
    tasks.register(copyTaskName) {
        copy {
            from "${buildDir}/../../../Resources"
            exclude "**/*.gz"
            exclude "**/*.pem"
            into "${buildDir}/intermediates/assets/${variant.dirName}"
        }
    }
    tasks.getByName("pre${targetName}Build").dependsOn copyTaskName
}
```

The problem is that the ".pem" files are still copies across to the output folder, and are embedded in the APK.  Is there something I'm missing in the script? The ".gz" files aren't actually excluded either, because the build process automatically extracts the contents into the output, which is why the ".gz" files do not appear in the output folder.

Now, if I comment out that copy task, the resources are still somehow copied during the build process, and I can't seem to track down where this is happening.  Does anyone have any idea on where this occurs in the build scripts? 

The only other way I've managed to get it to exclude files is with the following:
```
aaptOptions {
    ignoreAssetsPattern '!*.pem:'
}
```

I would prefer not to use the `aaptOptions` method since there doesn't seem to be a way to restrict it to release builds, so it activates in debug builds too.

**Comment by qoozta** [comment]
Maybe you can try copy the resources to somewhere else first to confirm the exclude is not working?
If the exclude is working, I think your cmake file has some copy task e.g. ax_copy_target_res
Also I suggest you use the assetPack which more convenient to publish on Google Play.

**Comment by rh101** [comment]
@qoozta Thanks for the suggestion, because simply testing a separate asset folder indicated that it was in fact doing the asset copy somewhere else, and actually doing it twice as a result of that copy task.

I tracked it down to here in `app/build.gradle`:

```
    sourceSets.main {
        java.srcDir "src"
        res.srcDir "res"
        manifest.srcFile "AndroidManifest.xml"
        assets.srcDir "../../Resources"  // <= this is the problem, as this results in all assets within the Resources folder being compied
    }
```

The task that is set below in that same file was copying them a second time, which now makes no sense at all:
```
android.applicationVariants.all { variant ->
    def project_root_folder = "${projectDir}/../.."
    def dest_assets_folder = "${projectDir}/assets"

    // delete previous files first
    delete dest_assets_folder
    def targetName = variant.name.capitalize()
    def copyTaskName = "copy${targetName}ResourcesToAssets"
    
    tasks.register(copyTaskName) {
        copy {
            from "${buildDir}/../../../Resources"
            into "${buildDir}/intermediates/assets/${variant.dirName}"
            exclude "**/*.gz"
        }
    }
    tasks.getByName("pre${targetName}Build").dependsOn copyTaskName
}
```

Also, that task has redundant code, such as the:
`def dest_assets_folder = "${projectDir}/assets"`
and 
`delete dest_assets_folder`

That doesn't do anything, since that directory is never used for assets, so it's deleting a directory that doesn't exist.

Now, if we comment out the `assets.srcDir "../../Resources"`, and let the copy task handle all the copying with correct exclusions etc., then any `.gz` files will no longer be automatically extracted in the build process, which may not be what the developer requires.  Source sets do not currently support the exclude/include patterns, so this functionality must be carried out in a custom task.  That limitation has already been logged with Google [here](https://issuetracker.google.com/issues/36988285).

For my specific case, commenting it out `assets.srcDir "../../Resources"` resolves the issue, so I just let the copy tasks handle the copying of the assets.

Regarding your suggestion to using the assetPack option, I'll look into it, since it seems beneficial, as long as it doesn't require any significant custom code to support it versus other platforms (such as iOS etc).

**Comment by qoozta** [comment]
Maybe your project gradle file is from Cocos2d-x 3.0 (NDK build) and not fully migrated to CMake build, axmol cpp-tests use the same copy task, however it will point the asset.srcDir to "assets", so it should work.

**Comment by rh101** [comment]
It's definitely completely integrated with cmake, and has been since cocos2d-x 3.17, with no trace of NDK build in the gradle files.  

It seems that there is a big difference between the cpp-tests 'gradle.build' and the cpp template 'gradle.build' files, particularly in the sections related to sourceSet and the copy task:

https://github.com/axmolengine/axmol/blob/dev/tests/cpp-tests/proj.android/app/build.gradle
vs
https://github.com/axmolengine/axmol/blob/dev/templates/cpp-template-default/proj.android/app/build.gradle

In cpp-tests, `asset.srcDir` is set to "assets", since the copy task initially copies all the resources to "assets", with exclusions etc.  It seems that the templates `gradle.build` file was updated at some point, but cpp-tests wasn't, and this is all going back to cocos2d-x, as it's not something that was introduced in axmol.
